controller/controller_server/csi_controller_server.py

```diff
@@ -153,70 +156,64 @@ def CreateVolume(self, request, context):  # pylint: disable=too-many-branches
             context.set_details('an internal exception occurred : {}'.format(ex))
             return csi_pb2.CreateVolumeResponse()
 
-    def _copy_to_existing_volume_from_snapshot(self, vol, src_snapshot_id, min_vol_size, array_mediator, pool):
-        vol_name = vol.volume_name
+    def _copy_to_existing_volume_from_source_object(self, volume, source_object_id, source_object_type,
+                                                    minimum_volume_size, array_mediator, pool):
+        volume_name = volume.name
         try:
-            src_snapshot = array_mediator.get_snapshot_by_id(src_snapshot_id)
-            if not src_snapshot:
-                raise controller_errors.SnapshotNotFoundError(src_snapshot_id)
-            src_snapshot_name = src_snapshot.snapshot_name
-            src_snapshot_capacity = src_snapshot.capacity_bytes
-            logger.debug("Copy snapshot {0} data to volume {1}.".format(src_snapshot_id, vol_name))
-            array_mediator.copy_to_existing_volume_from_snapshot(vol_name, src_snapshot_name, src_snapshot_capacity,
-                                                                 min_vol_size, pool)
-            logger.debug("Copy volume from snapshot finished")
-        except controller_errors.VolumeNotFoundError as ex:
-            logger.error("Volume not found while copying snapshot data to volume")
+            source_object = array_mediator.get_object_by_id(source_object_id, source_object_type)
```

Comment:
the type passed here should be `OBJECT_TYPE_NAME_*`, and not `VOLUME_SOURCE_*` (unless you want to unify them)

Reply:
what do you recommend?

Reply:
done

Reply:
you are still passing them, and these 4 variables are not unified into 2...
I don't see a reason to keep 4 if they have the same values by definition.

Reply:
need to pass.. I will unify to 2..

Reply:
done